mt-kahypar/partition/deep_multilevel.cpp

```diff
@@ -36,346 +36,777 @@
-    d_context.partition.max_part_weights.assign(d_context.partition.k, 0);
-    for (PartitionID part = 0; part < context.partition.k; ++part) {
-      d_context.partition.perfect_balance_part_weights[part / 2] +=
-        context.partition.perfect_balance_part_weights[part];
+
+  PartitionID get_maximum_number_of_blocks(const HypernodeID current_num_nodes) const {
+    const int num_levels = _desired_blocks.size();
+    for ( int i = num_levels - 1; i >= 0; --i ) {
+      const PartitionID k = _desired_blocks[i].size();
+      if ( current_num_nodes >= k * _contraction_limit_multiplier ) {
+        return k;
+      }
     }
+    return _desired_blocks.back().size();
+  }
 
-    d_context.partition.epsilon = info.computeAdaptiveEpsilon(d_context.partition.k);
-    for (PartitionID part = 0; part < d_context.partition.k; ++part) {
-      d_context.partition.max_part_weights[part] =
-        std::ceil(( 1.0 + d_context.partition.epsilon ) * d_context.partition.perfect_balance_part_weights[part]);
+  void printRBTree() const {
+    for ( size_t level = 0; level < _desired_blocks.size(); ++level ) {
+      std::cout << "Level " << (level + 1) << std::endl;
+      for ( size_t i = 0; i <  _desired_blocks[level].size(); ++i) {
+        std::cout << "(" << _desired_blocks[level][i]
+                  << ", [" << _target_blocks[level][i] << "," << _target_blocks[level][i + 1] << "]"
+                  << ", " << _perfectly_balanced_weights[level][i]
+                  << ", " << _max_part_weights[level][i] << ") ";
+      }
+      std::cout << std::endl;
     }
   }
 
-  // Coarsening Parameters
-  d_context.coarsening.contraction_limit = std::max(
-          d_context.partition.k * d_context.coarsening.contraction_limit_multiplier,
-          2 * ID(d_context.shared_memory.num_threads) *
-          d_context.coarsening.contraction_limit_multiplier);
-  d_context.setupMaximumAllowedNodeWeight(hypergraph.totalWeight());
-  d_context.setupThreadsPerFlowSearch();
+ private:
+  void precomputeRBTree(const Context& context) {
+    auto add_block = [&](const PartitionID k) {
+      const PartitionID start = _target_blocks.back().back();
+      _desired_blocks.back().push_back(k);
+      _target_blocks.back().push_back(start + k);
+      const HypernodeWeight perfect_part_weight = std::accumulate(
+        context.partition.perfect_balance_part_weights.cbegin() + start,
+        context.partition.perfect_balance_part_weights.cbegin() + start + k, 0);
+      const HypernodeWeight max_part_weight = std::accumulate(
+        context.partition.max_part_weights.cbegin() + start,
+        context.partition.max_part_weights.cbegin() + start + k, 0);
+      _perfectly_balanced_weights.back().push_back(perfect_part_weight);
+      _max_part_weights.back().push_back(max_part_weight);
+    };
+
+    int cur_level = 0;
+    bool should_continue = true;
+    // Simulates recursive bipartitioning
```

Comment:
We discussed doing this during the actual recursion instead of precomputing. That way we could adapt to the current block balance at each recursion step. WDYT?

Reply:
As we discussed, this is not possible when using individual part weights.